File: proyecto.py
import tkinter as tk
from tkinter import messagebox
from datetime import datetime
import random

# Función para calcular la edad a partir de la fecha de nacimiento
def calcular_edad(fecha_nacimiento):
    try:
        dia, mes, anio = map(int, fecha_nacimiento.split('/'))
        fecha_nacimiento = datetime(anio, mes, dia)
    except ValueError:
        messagebox.showerror("Error", "Fecha de nacimiento no válida. Por favor ingrese en formato dd/mm/aaaa.")
        return None

    fecha_actual = datetime.today()
    edad = fecha_actual.year - fecha_nacimiento.year
    if (fecha_actual.month < mes) or (fecha_actual.month == mes and fecha_actual.day < dia):
        edad -= 1
    return edad

# Función para determinar el grupo de edad
def determinar_grupo_edad(edad):
    if edad <= 1:
        return "Neonato"
    elif edad <= 12:
        return "Pediátrico"
    elif edad <= 17:
        return "Adolescente"
    elif edad <= 64:
        return "Adulto"
    else:
        return "Adulto Mayor"

# Función para iniciar la simulación
def iniciar_simulacion():
    nombre = entry_nombre.get()
    fecha_nacimiento = entry_fecha.get()

    # Calcular edad
    edad = calcular_edad(fecha_nacimiento)
    if edad is None:
        return

    # Determinar el grupo de edad
    grupo_edad = determinar_grupo_edad(edad)

    # Calcular FCM y rango esperado
    fcm = 220 - edad
    rango_min = fcm * 0.50
    rango_max = fcm * 0.85

    # Mostrar resultados iniciales
    result_var.set(f"Nombre: {nombre}\nEdad: {edad} años\nGrupo de Edad: {grupo_edad}\n"
                   f"Frecuencia Cardíaca Máxima: {fcm} BPM\nRango Esperado: {int(rango_min)} - {int(rango_max)} BPM")

    # Iniciar la simulación
    historial = []
    minuto = 1
    taquicardia_detectada = False
    alerta_leve = int(rango_max)
    alerta_grave = int(fcm * 0.90)
    frecuencia_cardiaca = random.randint(int(rango_min), int(rango_max))
    complicaciones = "Ninguna"

    while not taquicardia_detectada:
        frecuencia_cardiaca += random.randint(1, 2)

        if frecuencia_cardiaca >= alerta_grave:
            complicaciones = "ALERTA GRAVE: Frecuencia cardiaca cerca del límite máximo."
        elif frecuencia_cardiaca >= alerta_leve:
            complicaciones = "ALERTA LEVE: Frecuencia cardiaca fuera del rango esperado."
        else:
            complicaciones = "Ninguna"

        if frecuencia_cardiaca > fcm:
            taquicardia_detectada = True
            complicaciones = "PACIENTE SUFRIÓ UNA TAQUICARDIA"

        # Guardar el minuto en el historial
        historial.append([minuto, frecuencia_cardiaca, complicaciones])
        minuto += 1

    # Mostrar historial de la simulación en el área de texto
    historial_text.delete(1.0, tk.END)
    historial_text.insert(tk.END, f"Simulación de Ejercicio\n")
    historial_text.insert(tk.END, f"{'Minuto':<8} {'Frecuencia Cardiaca':<20} {'Complicaciones'}\n")
    historial_text.insert(tk.END, "-"*50 + "\n")
    
    for minuto, frecuencia, complicacion in historial:
        historial_text.insert(tk.END, f"{minuto:<8} {frecuencia} BPM   {complicacion}\n")

    historial_text.insert(tk.END, "\nSimulación Finalizada. ¡Gracias!")

# Configuración de la ventana principal
ventana = tk.Tk()
ventana.title("Simulación de Ejercicio")

# Etiquetas y entradas para los datos del paciente
tk.Label(ventana, text="Nombre del paciente:").grid(row=0, column=0, padx=10, pady=10, sticky="w")
entry_nombre = tk.Entry(ventana, width=30)
entry_nombre.grid(row=0, column=1, padx=10, pady=10)

tk.Label(ventana, text="Fecha de nacimiento (dd/mm/aaaa):").grid(row=1, column=0, padx=10, pady=10, sticky="w")
entry_fecha = tk.Entry(ventana, width=15)
entry_fecha.grid(row=1, column=1, padx=10, pady=10)

# Botón para iniciar la simulación
boton_simulacion = tk.Button(ventana, text="Iniciar Simulación de Ejercicio", command=iniciar_simulacion)
boton_simulacion.grid(row=2, column=0, columnspan=4, padx=10, pady=20)

# Etiqueta para mostrar los resultados
result_var = tk.StringVar()
tk.Label(ventana, textvariable=result_var, justify=tk.LEFT, width=60, anchor="w", relief="solid", padx=10, pady=10).grid(row=3, column=0, columnspan=4, padx=10, pady=10)

# Área de texto para mostrar el historial de la simulación
historial_text = tk.Text(ventana, width=70, height=15, wrap=tk.WORD)
historial_text.grid(row=4, column=0, columnspan=4, padx=10, pady=10)

# Ejecutar la ventana
ventana.mainloop()

File: proyecto.java

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class proyecto {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Ingreso de datos personales
        System.out.print("Ingrese el nombre del paciente: ");
        String nombre = sc.nextLine();

        System.out.print("Ingrese su fecha de nacimiento (dd mm aaaa): ");
        int dia = sc.nextInt();
        int mes = sc.nextInt();
        int anio = sc.nextInt();

        // Calcular edad
        int edad = calcularEdad(anio, mes, dia);
        
        // Determinar grupo de edad
        String grupoEdad = determinarGrupoEdad(edad);

        // Calcular FCM y Rango Esperado
        int fcm = 220 - edad;
        double rangoMin = fcm * 0.50;
        double rangoMax = fcm * 0.85;

        // Mostrar datos iniciales
        System.out.printf("\nNombre              : %s\n", nombre);
        System.out.printf("Edad                : %d years\n", edad);
        System.out.printf("Grupo de Edad       : %s\n", grupoEdad);
        System.out.printf("Rango esperado      : %d - %d BPM\n", (int)rangoMin, (int)rangoMax);
        System.out.printf("Frecuencia maxima   : %d BPM\n\n", fcm);

        // Esperar que el usuario presione Enter para comenzar la simulación
        System.out.println("Presione Enter para comenzar la simulacion del ejercicio...");
        sc.nextLine(); // Pausar hasta que el usuario presione Enter
        sc.nextLine(); // Otra vez para iniciar

        // Simulación del ejercicio
        simulacionEjercicio(fcm, rangoMin, rangoMax);
    }

    // Metodo para calcular la edad
    public static int calcularEdad(int anio, int mes, int dia) {
        int anioActual = 2024; 
        int mesActual = 11;    
        int diaActual = 18;    

        int edad = anioActual - anio;
        if (mesActual < mes || (mesActual == mes && diaActual < dia)) {
            edad--;
        }
        return edad;
    }

    // Metodo para determinar el grupo de edad
    public static String determinarGrupoEdad(int edad) {
        if (edad <= 1) {
            return "Neonato";
        } else if (edad <= 12) {
            return "Pediatrico";
        } else if (edad <= 17) {
            return "Adolescente";
        } else if (edad <= 64) {
            return "Adulto (18-64 years)";
        } else {
            return "Adulto mayor";
        }
    }

    // Metodo para simular el ejercicio
    public static void simulacionEjercicio(int fcm, double rangoMin, double rangoMax) {
        Scanner sc = new Scanner(System.in);
        int minuto = 1;
        boolean continuar = true;

        // Crear lista para almacenar los resultados
        List<String> historial = new ArrayList<>();

        // Umbrales para las alertas
        int alertaLeve = (int)rangoMax; // Alerta leve cuando se alcanza el rango maximo
        int alertaGrave = (int)(fcm * 0.90); // Alerta grave cuando se alcanza el 90% de la frecuencia maxima

        // Iniciar con un valor dentro del rango esperado
        int frecuenciaCardiaca = (int)(rangoMin + Math.random() * (rangoMax - rangoMin));
        String complicaciones = "Ninguna";

        // Encabezado de la tabla
        System.out.printf("\n%-8s %-25s %-12s %-50s %-25s\n", "Minuto", "Frecuencia cardiaca", "Duracion (min)", "Complicaciones", "Taquicardia");
        System.out.println("----------------------------------------------------------------------------------");

        // Bucle de simulacion de ejercicio
        boolean taquicardiaDetectada = false; // Variable para detectar si hubo taquicardia
        while (continuar) {
            // Incrementar la frecuencia de manera mas gradual (1 a 2 BPM por minuto)
            frecuenciaCardiaca += (int)(Math.random() * 2 + 1);

            // Detectar alertas segun los rangos
            if (frecuenciaCardiaca >= alertaGrave) {
                complicaciones = "ALERTA GRAVE: Frecuencia cardiaca cerca del limite maximo.";
            } else if (frecuenciaCardiaca >= alertaLeve) {
                complicaciones = "ALERTA LEVE: Frecuencia cardiaca fuera del rango esperado.";
            } else {
                complicaciones = "Ninguna";
            }

            // Si la frecuencia supera el rango maximo, detener la simulacion y dar mensaje de taquicardia
            if (frecuenciaCardiaca > fcm) {
                System.out.printf("\nEl paciente ha sufrido una taquicardia. Simulacion terminada.\n");
                complicaciones = "PACIENTE SUFRIO UNA TAQUICARDIA";
                taquicardiaDetectada = true;
                break;
            }

            // Almacenar los resultados del minuto
            historial.add(String.format("%-8d %-25s %-12d %-50s %-25s", minuto, frecuenciaCardiaca + " BPM", minuto, complicaciones, taquicardiaDetectada ? "Si" : "No"));
            // Mostrar los resultados del minuto
            System.out.printf("%-8d %-25s %-12d %-50s %-25s\n", minuto, frecuenciaCardiaca + " BPM", minuto, complicaciones, taquicardiaDetectada ? "Si" : "No");

            // Preguntar si continuar
            System.out.println("\nPresione Enter para continuar o cualquier otra tecla para finalizar...");
            String input = sc.nextLine();
            if (!input.equals("")) {
                continuar = false; // Finalizar simulacion si no se presiona Enter
            }
            
            minuto++;
        }

        // Mostrar la tabla completa con todos los resultados
        System.out.println("\nTabla de resultados finales:");
        System.out.println("Minuto    Frecuencia cardiaca      Duracion (min)        Complicaciones                 Taquicardia");
        System.out.println("----------------------------------------------------------------------------------------------");

        // Imprimir todas las filas almacenadas en el historial con el formato de tabla
        for (String linea : historial) {
            System.out.println(linea);
        }

        // Mostrar mensaje final
        System.out.println("\nPaciente entro en Taquicardia. Ejercicio finalizado. ¡Gracias!");
    }
}
